doc/api/async_hooks.md

```diff
@@ -682,6 +682,172 @@ never be called.
+everything that is executed within the callback of `run` (including further
+asynchronous operations) will have access to the store. The `exit` method has
+a similar behavior to stop context propagation.
+
+If an instance of AsyncContext is used for error management (for instance with
+`process.setUncaughtExceptionCaptureCallback`). Only the exceptions thrown
+in the scope of the callback function will be associated with the context.
+
+These methods are the safest to call as they provide strong scoping.
+
+#### When to choose `runAndReturn/exitAndReturn`
+These methods are synchronous and scoped. The callback function will be executed
+synchronously and it's return value will be returned by `runAndReturn`
+and `exitAndReturn`. The store will only be accessible from within the callback
+function and the asynchronous operations created within this scope.
+If the callback throws an error, `runAndReturn` and `exitAndReturn` will throw
+it too.
+
+These methods provide good scoping while being synchronous.
+
+#### What are the risks of `enterSync/exitSync`
+These methods are to be used only if you know what you are doing. After calling
+`enterSync`, the context will be propagated to everything that follows.
+For instance, in the case of an event emitter, if the first executed listener
+calls `enterSync`, the following ones will have access to the store.
+```js
+emitter.on('event', () => {
+    asyncContext.enterSync();
+});
+emitter.on('event', () => {
+    asyncContext.getStore(); // will return  the store.
+});
+```
+If an instance of AsyncContext is used for error management (for instance with
+`process.setUncaughtExceptionCaptureCallback`). All exceptions will be
+associated with the context.
+
+These methods are less safe than `run/exit` and `runAndReturn/exitAndReturn` as
+they have no scope.
+
+### new AsyncContext()
+
+* Returns: {Object}
+
+Creates a new instance of `AsyncContext`. Storage is only provided inside a `run`
+ method call.
+
+### asyncContext.run(callback)
+
+* `callback` {Function}
+  * `store` {Map}
+
+Calling `asyncContext.enter(callback)` will create a new asynchronous context.
+This method will call the callback with a `Map` known as the store as argument.
+The callback will be ran asynchronously.
+
+This store will be persistent through the following asynchronous calls.
+
+Currently there is a known limitation with this method. If called within an
+`async function` the context will be lost after the next `await` call.
+
+### asyncContext.exit(callback)
+
+* `callback` {Function}
+
+Calling `asyncContext.exit(callback)` will exit the context.
+Asynchronous operations following the callback cannot access the
+ store.
+The callback will be ran asynchronously.
+
+Currently there is a known limitation with this method. If called within an
+`async function` the context will be still available after the next `await`
+call.
+
+### asyncContext.runAndReturn(callback)
```

Comment:
Consider to allow passing arguments to callback similar as in `domain.run()`

Reply:
make sense. Will add that.